devtools/datasette/publish.py

```diff
@@ -82,9 +83,78 @@ def inspect_data(datasets: list[str], pudl_out: Path) -> str:
     return inspect_output
 
 
-def metadata(pudl_out: Path) -> str:
+def check_tables_have_metadata(
+    metadata: DatasetteMetadata,
+    pudl_out: Path,
+    datasets: list[str],
+) -> None:
+    """Check to make sure all tables in the databases have metadata.
+
+    This function fails if there are tables present in a database
+    that do not have structure metadata.
+
+    Args:
+        metadata: The structure metadata for the datasette deployment
+        pudl_out: The directory that contains the pudl outputs
+        datasets: The list of datasets to test.
+    """
+    resources = {}
+    resources |= metadata.xbrl_resources
+    resources["pudl"] = metadata.resources
+
+    resource_exceptions = {"pudl": {"alembic_version"}}
+
+    tables_missing_metadata_results = {}
+
+    for dataset in datasets:
+        database_path = pudl_out / dataset
+        dataset_name = database_path.stem
+        connection = sqlite3.connect(database_path)
+
+        cursor = connection.cursor()
+        query = "SELECT name FROM sqlite_master WHERE type='table';"
+        cursor.execute(query)
+
+        dataset_exceptions = resource_exceptions.get(dataset_name)
+
+        resources_in_database = {table_name[0] for table_name in cursor.fetchall()}
+        if dataset_exceptions:
+            resources_in_database = resources_in_database - dataset_exceptions
+        resources_with_metadata = {
+            resource.name for resource in resources[dataset_name]
+        }
+
+        cursor.close()
+        connection.close()
```

Comment:
Can we use a context manager instead of needing to close these connections?

Reply:
Yes there is a clean way to grab all tables using sqlalchemy.